[Site Isolation] Fix media/video-pause-play-resolve.html, media/video-pause-immediately.html and media/media-h264-webm-source.html

charliewolfe opened the pull request:
#### a13b3bb3bd132b3a90bf630f8e6f4bebc60fc800
<pre>
[Site Isolation] Fix media/video-pause-play-resolve.html, media/video-pause-immediately.html and media/media-h264-webm-source.html
<a href="https://bugs.webkit.org/show_bug.cgi?id=293078">https://bugs.webkit.org/show_bug.cgi?id=293078</a>
<a href="https://rdar.apple.com/151422482">rdar://151422482</a>

Reviewed by Alex Christensen.

Correctly update autoplayPolicy for iframe processes and begin using it whenever it&apos;s accessed from the
top document.

* LayoutTests/platform/mac-site-isolation/TestExpectations:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/webaudio/AudioContext.cpp:
(WebCore::AudioContext::constructCommon):
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::m_remote):
(WebCore::HTMLMediaElement::updateRateChangeRestrictions):
* Source/WebCore/loader/DocumentLoader.h:
* Source/WebCore/page/AutoplayPolicy.h: Copied from Source/WebKit/Shared/WebsiteAutoplayPolicy.h.
* Source/WebCore/page/Frame.h:
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::autoplayPolicy const):
* Source/WebCore/page/LocalFrame.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::requiresUserGestureForAudioPlayback const):
(WebCore::Page::requiresUserGestureForVideoPlayback const):
* Source/WebCore/page/Page.h:
* Source/WebCore/page/RemoteFrame.cpp:
(WebCore::m_autoplayPolicy):
(WebCore::RemoteFrame::autoplayPolicy const):
(WebCore::m_layerHostingContextIdentifier): Deleted.
* Source/WebCore/page/RemoteFrame.h:
* Source/WebKit/Shared/WebsiteAutoplayPolicy.cpp: Copied from Source/WebKit/Shared/WebsiteAutoplayPolicy.h.
(WebKit::core):
* Source/WebKit/Shared/WebsiteAutoplayPolicy.h:
* Source/WebKit/Shared/WebsitePoliciesData.cpp:
(WebKit::WebsitePoliciesData::applyToDocumentLoader):
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::applyWebsitePolicies):

Canonical link: <a href="https://commits.webkit.org/295041@main">https://commits.webkit.org/295041@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9120b81aa508eb59ddfd29d4affaa9680d5be4fe

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103844 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23546 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13868 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109037 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54500 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105884 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23899 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32092 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78899 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106850 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18559 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93677 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59228 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18358 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11727 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53872 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/88119 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11786 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111424 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31000 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22859 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87911 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31362 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89878 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87563 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32442 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10169 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25365 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16866 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30929 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36239 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30722 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34059 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32283 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Shared/WebsiteAutoplayPolicy.cpp:
```diff
@@ -0,0 +1,49 @@
+        return WebCore::AutoplayPolicy::Deny;
+    }
+    ASSERT_NOT_REACHED();
+    return { };
```
WebCore::AutoplayPolicy::Default

achristensen07 commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1385,6 +1385,13 @@ OptionSet<AdvancedPrivacyProtections> LocalFrame::advancedPrivacyProtections() c
+{
+    if (RefPtr documentLoader = loader().activeDocumentLoader())
+        return documentLoader->autoplayPolicy();
+    return { };
```
AutoplayPolicy::Default

achristensen07 commented on Source/WebCore/loader/DocumentLoader.h:
```diff
@@ -30,6 +30,7 @@
 #pragma once
 
 #include "AdvancedPrivacyProtections.h"
+#include "AutoplayPolicy.h"
```
Can this be a forward declaration instead?

charliewolfe replied:
The default initialization below uses AutoplayPolicy::Default

charliewolfe committed:
[Site Isolation] Fix media/video-pause-play-resolve.html, media/video-pause-immediately.html and media/media-h264-webm-source.html
https://bugs.webkit.org/show_bug.cgi?id=293078
rdar://151422482

Reviewed by Alex Christensen.

Correctly update autoplayPolicy for iframe processes and begin using it whenever it's accessed from the
top document.

* LayoutTests/platform/mac-site-isolation/TestExpectations:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/webaudio/AudioContext.cpp:
(WebCore::AudioContext::constructCommon):
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::m_remote):
(WebCore::HTMLMediaElement::updateRateChangeRestrictions):
* Source/WebCore/loader/DocumentLoader.h:
* Source/WebCore/page/AutoplayPolicy.h: Copied from Source/WebKit/Shared/WebsiteAutoplayPolicy.h.
* Source/WebCore/page/Frame.h:
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::autoplayPolicy const):
* Source/WebCore/page/LocalFrame.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::requiresUserGestureForAudioPlayback const):
(WebCore::Page::requiresUserGestureForVideoPlayback const):
* Source/WebCore/page/Page.h:
* Source/WebCore/page/RemoteFrame.cpp:
(WebCore::m_autoplayPolicy):
(WebCore::RemoteFrame::autoplayPolicy const):
(WebCore::m_layerHostingContextIdentifier): Deleted.
* Source/WebCore/page/RemoteFrame.h:
* Source/WebKit/Shared/WebsiteAutoplayPolicy.cpp: Copied from Source/WebKit/Shared/WebsiteAutoplayPolicy.h.
(WebKit::core):
* Source/WebKit/Shared/WebsiteAutoplayPolicy.h:
* Source/WebKit/Shared/WebsitePoliciesData.cpp:
(WebKit::WebsitePoliciesData::applyToDocumentLoader):
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::applyWebsitePolicies):

Canonical link: https://commits.webkit.org/295041@main
